File: contests/lc/120.Triangle/triangle.go
// Package triangle implements helpers for working with the classic
// "Triangle" minimum-path-sum dynamic programming problem.
package triangle

// minimumTotal returns the smallest possible sum of a path that starts at the
// apex of the triangle and, moving only to adjacent numbers on the row below,
// reaches the base. The implementation runs bottom-up, reusing the last row as
// a rolling buffer so it achieves O(n^2) time with O(n) additional space.
func minimumTotal(triangle [][]int) int {
	n := len(triangle)
	f := make([]int, n+1)
	for i := n - 1; i >= 0; i-- {
		for j := 0; j <= i; j++ {
			f[j] = min(f[j], f[j+1]) + triangle[i][j]
		}
	}
	return f[0]
}

File: contests/lc/120.Triangle/triangle.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if not triangle:
            return 0

        dp = triangle[-1][:]
        for i in range(len(triangle) - 2, -1, -1):
            row = triangle[i]
            for j in range(len(row)):
                dp[j] = min(dp[j], dp[j + 1]) + row[j]
        return dp[0]
